app/models/user.rb

```diff
@@ -595,15 +627,48 @@ def create_personal_herbarium
   # Project that the User is a member of.
   def all_editable_species_lists(include: nil)
     @all_editable_species_lists ||= begin
-      results = species_lists
       if projects_member.any?
-        project_ids = projects_member.map(&:id).join(",")
-        results += SpeciesList.includes(include).find_by_sql(%(
-          SELECT species_lists.* FROM species_lists, projects_species_lists
-          WHERE species_lists.user_id != #{id}
-            AND projects_species_lists.project_id IN (#{project_ids})
-            AND projects_species_lists.species_list_id = species_lists.id
-        ))
+        # project_ids = projects_member.map(&:id).join(",")
+        # results = species_lists
+        # results += SpeciesList.includes(include).find_by_sql(%(
+        #   SELECT species_lists.* FROM species_lists, projects_species_lists
+        #   WHERE species_lists.user_id != #{id}
+        #     AND projects_species_lists.project_id IN (#{project_ids})
+        #     AND projects_species_lists.species_list_id = species_lists.id
+        #   ))
+        # puts("original SQL")
+        # puts(results.uniq.sort.pluck(:id).inspect)
+
+        # Revised SQL: user species_lists are in WHERE, avoiding extra load
+        # project_ids = projects_member.map(&:id).join(",")
+        # results = SpeciesList.includes(include).find_by_sql(%(
+        #   SELECT species_lists.* FROM species_lists, projects_species_lists
+        #   WHERE (species_lists.user_id = #{id} )
+        #   OR (species_lists.user_id != #{id}
+        #     AND projects_species_lists.project_id IN (#{project_ids})
+        #     AND projects_species_lists.species_list_id = species_lists.id)
+        #   ))
+        # puts("revised SQL")
+        # puts(results.uniq.sort.pluck(:id).inspect)
+
+        project_ids = projects_member.map(&:id)
+        sl = SpeciesList.arel_table
+        psl = Arel::Table.new(:projects_species_lists)
+        constraints = sl.create_on(
+          sl[:user_id].eq(id).or(
+            sl[:user_id].not_eq(id).and(
```

Comment:
sl[:user_id].not_eq(id) is unnecessary

But I still don't think this query is right.  I guess I'll try writing a test to demonstrate my case.

Reply:
Note that the old method starts by setting `results = species_lists` at the outset, **outside** the conditional, and then, "if projects_member.any?" **adds** the new results ("not user's own, but in a project they're a member of") to those results. That condition resulted in two separate loads of `species_lists`, that's why i put both `user's own, or not own, but member of` in the same query here. It consolidates to a single load, and then the 'if not any?' condition sets `results = species_lists`, the old default.

This is the query i spent the most time on, but now that i think about it since it's an inner join maybe you're right.  For "if projects_member.any?", my logic is "either (own species lists) or (not own, but in a project they're in)". I tried it many ways and iirc the `sl[:user_id].not_eq(id)` was necessary to get the results, but now that you mention it the logic seems like it should be "either (own species lists) or (in a project they're in)". i will try it again now.

Reply:
Welp, you're definitely right that `sl[:user_id].not_eq(id)` was unnecessary. At least according to the spot-check tests i'm doing, on the db snapshot. I'll commit that change.

Reply:
Or maybe i should wait til you've devised the test...